File: code_100/java/com/interview/tree/ConstructTreeFromInOrderPreOrder.java
package com.interview.tree;

/**
 * http://www.geeksforgeeks.org/construct-tree-from-given-inorder-and-preorder-traversal/
 * Test cases:
 * Empty tree
 * One node tree
 * All left side tree
 * All right side tree
 * Mixed tree
 * Full tree
 * complete tree
 */
public class ConstructTreeFromInOrderPreOrder {
    
    private int index = 0;
    public Node createTree(int inorder[],int preorder[]){
        Node result =  createTree(inorder,preorder,0,inorder.length-1);
        index = 0;
        return result;
    }
    
    private Node createTree(int inorder[],int preorder[], int start, int end){
        if(start > end){
            return null;
        }
        int i;
        for(i=start; i <= end; i++){
            if(preorder[index] == inorder[i]){
                break;
            }
        }
        Node node = Node.newNode(preorder[index]);
        index++;
        node.left = createTree(inorder,preorder,start,i-1);
        node.right = createTree(inorder,preorder,i+1,end);
        return node;
    }
}

File: showmecode_100/python/tree/construct_tree_from_inorder_preorder.py
from binary_tree import Node


class ConstructTreeFromInorderPreOrder:
    def __init__(self):
        self.index = 0

    def _createTree(self, inorder, preorder, start, end):
        if start > end:
            return None
        i = 0
        for i in range(start, end + 1):
            if preorder[self.index] == inorder[i]:
                break

        node = Node.newNode(preorder[self.index])
        self.index += 1
        node.left = self._createTree(inorder, preorder, start, i - 1)
        node.right = self._createTree(inorder, preorder, i + 1, end)
        return node

    def createTree(self, inorder, preorder):
        return self._createTree(inorder, preorder, 0, len(inorder) - 1)
